### PR TITLE
Add GroupedSPMe model

### PR DESCRIPTION
**Pull request by NicolaCourtier**
# Description

Add a grouped parameter version of the SPMe for parameter identification.

## Issue reference
New feature

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [x] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [x] Examples: A change to existing or additional examples.
- [ ] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [x] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

**Comment by NicolaCourtier** [comment]
This implementation works with input currents and EIS, but it can only reliably run an experiment without the double layer (i.e. not with `{"surface form": "differential"}`). We are not currently sure whether this issue is on the PyBOP or PyBaMM side.

File: examples/scripts/comparison_examples/grouped_SPMe.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            ax.plot(np.real(impedances[:, ii]), -np.imag(impedances[:, ii]), "r--")
+
+# Show figure
+ax.set(xlabel=r"$Z_r(\omega)$ [$\Omega$]", ylabel=r"$-Z_j(\omega)$ [$\Omega$]")
+ax.set_aspect("equal", "box")
+plt.show()
+fig.savefig("Nyquist.png")
+
+# mdic = {"Z": impedances, "f": frequencies, "SOC": SOCs}
+# savemat("Simulated data SPMe/Z_SPMe_SOC_Pybop_chen2020.mat", mdic)
```
Comment (brosaplanella) [inline examples/scripts/comparison_examples/grouped_SPMe.py:120]: Need removing (or uncommenting)

File: examples/scripts/comparison_examples/grouped_SPMe.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ce0 = parameter_set["Initial concentration in electrolyte [mol.m-3]"]
+T = parameter_set["Ambient temperature [K]"]
+parameter_set["Electrolyte diffusivity [m2.s-1]"] = parameter_set[
+    "Electrolyte diffusivity [m2.s-1]"
+](ce0, T)
+parameter_set["Electrolyte conductivity [S.m-1]"] = parameter_set[
+    "Electrolyte conductivity [S.m-1]"
+](ce0, T)
+
+# Make the conductivities artificially large
```
Comment (brosaplanella) [inline examples/scripts/comparison_examples/grouped_SPMe.py:25]: Why is that?

File: pybop/models/lithium_ion/basic_SPMe.py [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+            "left": (pybamm.boundary_value(sto_e_n, "right"), "Dirichlet"),
+            "right": (pybamm.boundary_value(sto_e_p, "left"), "Dirichlet"),
+        }
+        self.boundary_conditions[sto_e_p] = {
+            "left": (
+                tau_e_p * pybamm.boundary_gradient(sto_e_sep, "right") / tau_e_sep,
+                "Neumann",
+            ),
+            "right": (Scalar(0), "Neumann"),
+        }
```
Comment (brosaplanella) [inline pybop/models/lithium_ion/basic_SPMe.py:315]: If variables are concatenated into `sto_e`, then we should be able to define an equation for `sto_e` and PyBaMM would deal with the continuity conditions.

File: pybop/models/lithium_ion/basic_SPMe.py [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+            "Separator electrolyte stoichiometry",
+            domain="separator",
+        )
+        sto_e_p = Variable(
+            "Positive electrode electrolyte stoichiometry",
+            domain="positive electrode",
+        )
+        # Concatenations combine several variables into a single variable, to simplify
+        # implementing equations that hold over several domains
+        # sto_e = pybamm.concatenation(sto_e_n, sto_e_sep, sto_e_p)
```
Comment (brosaplanella) [inline pybop/models/lithium_ion/basic_SPMe.py:102]: See comment below about concatenation

File: pybop/models/lithium_ion/basic_SPMe.py [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+            j_p = j0_p * (
+                pybamm.exp((1 - alpha) * eta_p / RT_F)
+                - pybamm.exp(-alpha * eta_p / RT_F)
+            )
+        else:
+            # Assuming alpha = 0.5
+            j_n = PrimaryBroadcast(I / (3 * Q_th_n), "negative electrode")
+            j_p = PrimaryBroadcast(-I / (3 * Q_th_p), "positive electrode")
+            eta_n = 2 * RT_F * pybamm.arcsinh(j_n / (2 * j0_n))
+            eta_p = 2 * RT_F * pybamm.arcsinh(j_p / (2 * j0_p))
```
Comment (brosaplanella) [inline pybop/models/lithium_ion/basic_SPMe.py:240]: I think these would be better as an `else` statement in the previous `if` statement, where `eta_n` and `eta_p` are defined for the double layer.

File: pybop/models/lithium_ion/basic_SPMe.py [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+            v_s_n = pybamm.x_average(eta_n + U_n)
+            v_s_p = pybamm.x_average(eta_p + U_p)
```
Comment (brosaplanella) [inline pybop/models/lithium_ion/basic_SPMe.py:242]: And those maybe in the next if statement, so they all include the if/else for the same variables

File: pybop/models/lithium_ion/basic_SPMe.py [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+        # # Set battery specific variables
+        # pybamm.logger.debug(f"Setting voltage variables ({self.name})")
+        # self.set_voltage_variables()
+
+        # pybamm.logger.debug(f"Setting SoC variables ({self.name})")
+        # self.set_soc_variables()
+
+        # pybamm.logger.debug(f"Setting degradation variables ({self.name})")
+        # self.set_degradation_variables()
+        # self.set_summary_variables()
```
Comment (brosaplanella) [inline pybop/models/lithium_ion/basic_SPMe.py:440]: Uncomment/remove

File: pybop/models/lithium_ion/echem.py
Comment (brosaplanella) [inline pybop/models/lithium_ion/echem.py:1]: I am a bit confused, why are `BaseGroupedSPMe` and `GroupedSPMe` different?

**Review by brosaplanella — COMMENTED** [review]
Looks good, just some comments!

File: examples/scripts/comparison_examples/grouped_SPMe.py (hunk 2 above)
Comment (NicolaCourtier) [reply examples/scripts/comparison_examples/grouped_SPMe.py:25]: For testing! Will remove

File: pybop/models/lithium_ion/echem.py
Comment (NicolaCourtier) [reply pybop/models/lithium_ion/echem.py:1]: This follows the same structure we have for other models, e.g. `BaseWeppnerHuggins` (the PyBaMM model) and `WeppnerHuggins` (the PyBOP model).

File: pybop/models/lithium_ion/basic_SPMe.py (hunk 5 above)
Comment (NicolaCourtier) [reply pybop/models/lithium_ion/basic_SPMe.py:219]: I've combined the double layer additions into one if statement under the double layer heading, is this better?

Comment (brosaplanella) [reply pybop/models/lithium_ion/basic_SPMe.py:219]: Yup, looks good

**Review by brosaplanella — APPROVED** [review]

**Pull request merged**